plugin exceptions

#1
java.lang.NullPointerException: Parameter specified as non-null is null: method org.digma.intellij.plugin.ui.common.EnvironmentUtilKt.getSortedEnvironments, parameter envsList
	at org.digma.intellij.plugin.ui.common.EnvironmentUtilKt.getSortedEnvironments(EnvironmentUtil.kt)
	at org.digma.intellij.plugin.toolwindow.recentactivity.DigmaBottomToolWindowFactory.processRecentActivityGetDataRequest(DigmaBottomToolWindowFactory.java:207)
	at org.digma.intellij.plugin.toolwindow.recentactivity.DigmaBottomToolWindowFactory$1.lambda$onQuery$0(DigmaBottomToolWindowFactory.java:132)
	at org.digma.intellij.plugin.common.AsynchronousBackgroundTask.run(AsynchronousBackgroundTask.kt:15)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:429)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsynchronously$6(CoreProgressManager.java:480)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:252)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:186)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:604)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:679)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:635)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:603)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:173)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:252)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)

#2

java.lang.NullPointerException: Cannot invoke "java.util.List.stream()" because the return value of "org.digma.intellij.plugin.analytics.AnalyticsService.getInsights(java.util.List)" is null
	at org.digma.intellij.plugin.document.DocumentInfoContainer.loadAllInsightsForCurrentDocument(DocumentInfoContainer.java:81)
	at org.digma.intellij.plugin.document.DocumentInfoContainer.update(DocumentInfoContainer.java:68)
	at org.digma.intellij.plugin.document.DocumentInfoService.addCodeObjects(DocumentInfoService.java:110)
	at org.digma.intellij.plugin.editor.EditorEventsHandler.lambda$selectionChanged$0(EditorEventsHandler.java:154)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:20)
	at com.intellij.util.RunnableCallable.call(RunnableCallable.java:11)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$OTelMonitor.callWrapped(NonBlockingReadActionImpl.java:746)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$OTelMonitor$MonitoredComputation.call(NonBlockingReadActionImpl.java:778)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.insideReadAction(NonBlockingReadActionImpl.java:573)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.lambda$attemptComputation$3(NonBlockingReadActionImpl.java:537)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1111)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runInReadActionWithWriteActionPriority$0(ProgressIndicatorUtils.java:71)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtilService.runActionAndCancelBeforeWrite(ProgressIndicatorUtilService.java:63)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runActionAndCancelBeforeWrite(ProgressIndicatorUtils.java:128)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runWithWriteActionPriority$1(ProgressIndicatorUtils.java:109)
	at com.intellij.openapi.progress.ProgressManager.lambda$runProcess$0(ProgressManager.java:68)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:186)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:604)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:679)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:635)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:603)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:173)
	at com.intellij.openapi.progress.ProgressManager.runProcess(ProgressManager.java:68)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runWithWriteActionPriority(ProgressIndicatorUtils.java:106)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runInReadActionWithWriteActionPriority(ProgressIndicatorUtils.java:71)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.attemptComputation(NonBlockingReadActionImpl.java:537)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.lambda$transferToBgThread$1(NonBlockingReadActionImpl.java:444)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:243)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$200(BoundedTaskExecutor.java:29)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.executeFirstTaskAndHelpQueue(BoundedTaskExecutor.java:216)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.run(BoundedTaskExecutor.java:208)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)

#3

java.lang.RuntimeException: Cannot invoke (class=, method=environmentChanged, topic=ENVIRONMENT_CHANGED_TOPIC)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:639)
	at com.intellij.util.messages.impl.MessageBusImplKt.deliverMessage(MessageBusImpl.kt:395)
	at com.intellij.util.messages.impl.MessageBusImplKt.pumpWaiting(MessageBusImpl.kt:374)
	at com.intellij.util.messages.impl.MessageBusImplKt.access$pumpWaiting(MessageBusImpl.kt:1)
	at com.intellij.util.messages.impl.MessagePublisher.invoke(MessageBusImpl.kt:433)
	at jdk.proxy6/jdk.proxy6.$Proxy246.environmentChanged(Unknown Source)
	at org.digma.intellij.plugin.toolwindow.recentactivity.DigmaBottomToolWindowFactory.lambda$createToolWindowContent$0(DigmaBottomToolWindowFactory.java:104)
	at org.digma.intellij.plugin.common.Backgroundable$1.run(Backgroundable.java:29)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:429)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsynchronously$6(CoreProgressManager.java:480)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:252)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:186)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:604)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:679)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:635)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:603)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:173)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:252)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.RuntimeException: Cannot invoke (class=EnvironmentChangeHandler, method=environmentChanged, topic=ENVIRONMENT_CHANGED_TOPIC)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:639)
	at com.intellij.util.messages.impl.MessageBusImplKt.deliverMessage(MessageBusImpl.kt:395)
	at com.intellij.util.messages.impl.MessageBusImplKt.pumpWaiting(MessageBusImpl.kt:364)
	at com.intellij.util.messages.impl.MessageBusImplKt.access$pumpWaiting(MessageBusImpl.kt:1)
	at com.intellij.util.messages.impl.MessagePublisher.invoke(MessageBusImpl.kt:433)
	at jdk.proxy6/jdk.proxy6.$Proxy247.usageStatusChanged(Unknown Source)
	at org.digma.intellij.plugin.analytics.AnalyticsService.notifyUsageStatusChanged(AnalyticsService.java:246)
	at org.digma.intellij.plugin.analytics.AnalyticsService.getUsageStatus(AnalyticsService.java:236)
	at org.digma.intellij.plugin.summary.SummariesProvider.getEnvironmentStatuses(SummariesProvider.java:52)
	at org.digma.intellij.plugin.ui.service.SummaryViewService.reload(SummaryViewService.kt:101)
	at org.digma.intellij.plugin.ui.service.SummaryViewService.reloadSummariesPanelInBackground$lambda$1(SummaryViewService.kt:89)
	at org.digma.intellij.plugin.common.Backgroundable.ensureBackground(Backgroundable.java:34)
	at org.digma.intellij.plugin.ui.service.SummaryViewService.reloadSummariesPanelInBackground(SummaryViewService.kt:95)
	at org.digma.intellij.plugin.ui.service.SummaryViewService.access$reloadSummariesPanelInBackground(SummaryViewService.kt:23)
	at org.digma.intellij.plugin.ui.service.SummaryViewService$2.environmentChanged(SummaryViewService.kt:61)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeMethod(MessageBusImpl.kt:655)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:621)
	... 27 more
Caused by: java.lang.RuntimeException: Cannot invoke (class=EnvironmentsDropdownPanel$$Lambda$6624/0x0000000302aab318, method=usageStatusChanged, topic=USAGE_STATUS_CHANGED_TOPIC)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:639)
	at com.intellij.util.messages.impl.MessageBusImplKt.deliverMessage(MessageBusImpl.kt:395)
	at com.intellij.util.messages.impl.MessageBusImplKt.pumpWaiting(MessageBusImpl.kt:374)
	at com.intellij.util.messages.impl.MessageBusImplKt.access$pumpWaiting(MessageBusImpl.kt:1)
	at com.intellij.util.messages.impl.MessagePublisher.invoke(MessageBusImpl.kt:427)
	at jdk.proxy6/jdk.proxy6.$Proxy247.usageStatusChanged(Unknown Source)
	at org.digma.intellij.plugin.analytics.AnalyticsService.notifyUsageStatusChanged(AnalyticsService.java:246)
	at org.digma.intellij.plugin.analytics.AnalyticsService.getUsageStatus(AnalyticsService.java:236)
	at org.digma.intellij.plugin.document.DocumentInfoContainer.loadAllInsightsForCurrentDocument(DocumentInfoContainer.java:93)
	at org.digma.intellij.plugin.document.DocumentInfoContainer.updateCache(DocumentInfoContainer.java:73)
	at org.digma.intellij.plugin.document.DocumentInfoService.lambda$updateCacheForAllOpenedDocuments$0(DocumentInfoService.java:81)
	at java.base/java.util.HashMap.forEach(HashMap.java:1421)
	at java.base/java.util.Collections$SynchronizedMap.forEach(Collections.java:2735)
	at org.digma.intellij.plugin.document.DocumentInfoService.updateCacheForAllOpenedDocuments(DocumentInfoService.java:81)
	at org.digma.intellij.plugin.document.DocumentInfoService.environmentChanged(DocumentInfoService.java:76)
	at org.digma.intellij.plugin.emvironment.EnvironmentChangeHandler.environmentChanged(EnvironmentChangeHandler.java:33)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeMethod(MessageBusImpl.kt:655)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:621)
	... 43 more
Caused by: java.lang.NullPointerException: newUsageStatus must not be null
	at org.digma.intellij.plugin.ui.common.EnvironmentsDropdownPanel._init_$lambda$0(EnvironmentsDropdownPanel.kt:67)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeMethod(MessageBusImpl.kt:655)
	at com.intellij.util.messages.impl.MessageBusImplKt.invokeListener(MessageBusImpl.kt:621)
	... 60 more


#4

Unhandled coroutine throwable

java.lang.NullPointerException: Cannot invoke "org.digma.intellij.plugin.model.rest.insights.CodeObjectInsightsStatusResponse.getCodeObjectsWithInsightsStatus()" because "response" is null
	at org.digma.intellij.plugin.insights.InsightsProvider.getInsightStatus(InsightsProvider.java:70)
	at org.digma.intellij.plugin.ui.service.InsightsViewService.fetchForInsightStatusAndUpdateUi(InsightsViewService.kt:88)
	at org.digma.intellij.plugin.ui.service.InsightsViewService$updateInsightsModel$3.invokeSuspend(InsightsViewService.kt:74)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)

--- a/ide-common/src/main/java/org/digma/intellij/plugin/analytics/AnalyticsService.java
+++ b/ide-common/src/main/java/org/digma/intellij/plugin/analytics/AnalyticsService.java
@@ -10,7 +10,9 @@ import org.digma.intellij.plugin.common.Backgroundable;
 import org.digma.intellij.plugin.common.CommonUtils;
 import org.digma.intellij.plugin.log.Log;
 import org.digma.intellij.plugin.model.InsightType;
+import org.digma.intellij.plugin.model.discovery.EndpointInfo;
 import org.digma.intellij.plugin.model.discovery.MethodInfo;
+import org.digma.intellij.plugin.model.discovery.SpanInfo;
 import org.digma.intellij.plugin.model.rest.debugger.DebuggerEventRequest;
 import org.digma.intellij.plugin.model.rest.errordetails.CodeObjectErrorDetails;
 import org.digma.intellij.plugin.model.rest.errors.CodeObjectError;
@@ -237,7 +239,7 @@ public class AnalyticsService implements Disposable {
     public CodeObjectInsightsStatusResponse getCodeObjectInsightStatus(List<MethodInfo> methodInfos) throws AnalyticsServiceException {
         var env = getCurrentEnvironment();
         var methodWithCodeObjects = methodInfos.stream()
-                .map(it -> toMethodWithCodeObjects(it))
+                .map(AnalyticsService::toMethodWithCodeObjects)
                 .toList();
         return executeCatching(() -> analyticsProviderProxy.getCodeObjectInsightStatus(new InsightOfMethodsRequest(env, methodWithCodeObjects)));
     }
@@ -245,8 +247,8 @@ public class AnalyticsService implements Disposable {
     @VisibleForTesting
     public static MethodWithCodeObjects toMethodWithCodeObjects(MethodInfo methodInfo) {
         return new MethodWithCodeObjects(methodInfo.idWithType(),
-                methodInfo.getSpans().stream().map(it -> it.idWithType()).toList(),
-                methodInfo.getEndpoints().stream().map(it -> it.idWithType()).toList()
+                methodInfo.getSpans().stream().map(SpanInfo::idWithType).toList(),
+                methodInfo.getEndpoints().stream().map(EndpointInfo::idWithType).toList()
         );
     }
 

--- a/ide-common/src/main/java/org/digma/intellij/plugin/insights/InsightsProvider.java
+++ b/ide-common/src/main/java/org/digma/intellij/plugin/insights/InsightsProvider.java
@@ -67,8 +67,12 @@ public class InsightsProvider {
     public InsightStatus getInsightStatus(@NotNull MethodInfo methodInfo) {
         try {
             CodeObjectInsightsStatusResponse response = analyticsService.getCodeObjectInsightStatus(List.of(methodInfo));
-            MethodWithInsightStatus methodResp = response.getCodeObjectsWithInsightsStatus().iterator().next();
-            return methodResp.getInsightStatus();
+            if (response != null) {
+                MethodWithInsightStatus methodResp = response.getCodeObjectsWithInsightsStatus().iterator().next();
+                return methodResp.getInsightStatus();
+            } else {
+                return InsightStatus.Unknown;
+            }
         } catch (AnalyticsServiceException e) {
             Log.log(LOGGER::debug, "AnalyticsServiceException for getCodeObjectInsightStatus for {}: {}", methodInfo.getId(), e.getMessage());
             return InsightStatus.Unknown;

--- a/ide-common/src/main/kotlin/org/digma/intellij/plugin/ui/service/InsightsViewService.kt
+++ b/ide-common/src/main/kotlin/org/digma/intellij/plugin/ui/service/InsightsViewService.kt
@@ -68,8 +68,7 @@ class InsightsViewService(project: Project) : AbstractViewService(project) {
 
             if (!insightsListContainer.listViewItems.isNullOrEmpty()) {
                 model.status = UiInsightStatus.InsightExist
-            }
-            else {
+            } else {
                 model.status = UiInsightStatus.Unknown
                 val newLifetimeDefinition = LifetimeDefinition()
                 newLifetimeDefinition.lifetime.launchBackground {

--- a/ide-common/src/main/kotlin/org/digma/intellij/plugin/ui/service/InsightsViewService.kt
+++ b/ide-common/src/main/kotlin/org/digma/intellij/plugin/ui/service/InsightsViewService.kt
@@ -79,8 +79,6 @@ class InsightsViewService(project: Project) : AbstractViewService(project) {
                     if (model.status == UiInsightStatus.Unknown && insightsListContainer.usageStatus.environmentStatuses.isNullOrEmpty()) {
                         model.status = UiInsightStatus.NoSpanData
                     }
-
-                    notifyModelChangedAndUpdateUi()
                 }
             }
 

